### ISSUE TITLE
No qrcode related to existing account

### ISSUE THREAD

**Issue by pleinad** [alfio-event/alf.io#379]
In the User page I cannot find my check-in operator qrcode to register on the check-in app. And in the iPhone verision of the app is not possible to enter the credentials manually.

**Comment by cbellone** [comment]
> In the User page I cannot find my check-in operator qrcode to register on the check-in app

The QR-Code includes the user password, in order to allow the mobile app to perform the login.

The user password can be displayed only after the user generation. After that we cannot decrypt it anymore. That's why you cannot find your check-in operator qrcode. 
We are trying to figure out more effective ways of doing it. Do you have any suggestions?

> And in the iPhone verision of the app is not possible to enter the credentials manually.

Actually the iPhone version of the app will be soon released on Android, so the "manual insert" functionality will be removed, because most of the people using it found it confusing.

Do you think that we should reintroduce it as "advanced option"?

**Comment by pleinad** [comment]
Hi Celestino, thank you for the answer!
> The QR-Code includes the user password, in order to allow the mobile app to perform the login.
> 
> The user password can be displayed only after the user generation. After that we cannot decrypt it anymore. That's why you cannot find your check-in operator qrcode.
> We are trying to figure out more effective ways of doing it. Do you have any suggestions?

Yep. You can store the image (the qrcode) in the profile of the user using a secret. For example you can use the password hash as seed for the encryption alghoritm. For sure the password hash is a secret known by the application about the user, but the user cannot read it... Everytime the password is changed, the encrypted image should be regenerated and saved. It could be an idea...
Alternatively (and easier) you can offer to the user the ability to download the qrcode telling him in clear manner that the qrcode will not be stored, so the user must save it in a secure place.

> Actually the iPhone version of the app will be soon released on Android, so the "manual insert" functionality will be removed, because most of the people found it confusing.
> 
> Do you think that we should reintroduce it as "advanced option"?

Next release on the Android store make the problem more relevant. I think that the scan as now is offered on the iPhone app is ok! But you should mark the qrcode as the operator check-in qrcode, making it different from another similar image. For example it could have a red border and a specific text...

My 2 cents. Thank you for your beautiful work!
Daniel.

**Comment by cbellone** [comment]
> Yep. You can store the image (the qrcode) in the profile of the user using a secret. For example you can use the password hash as seed for the encryption alghoritm. For sure the password hash is a secret known by the application about the user, but the user cannot read it... Everytime the password is changed, the encrypted image should be regenerated and saved. It could be an idea...

that would be clever indeed, but I am afraid it would also be insecure. We store the user's password using a [BCryptPasswordEncoder](https://docs.spring.io/spring-security/site/docs/4.0.4.RELEASE/apidocs/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.html), which means using an "one-way" encryption/hashing.
Now, if we save the same information inside an image and then we store it using a two-way encryption, we have also to store the key somewhere. If someone finds the key (it's not that difficult, since the code which saves that information will be public), he/she would have access to ALL the passwords stored, no matter how strong is the encryption. And we cannot allow that.

>Next release on the Android store make the problem more relevant. I think that the scan as now is offered on the iPhone app is ok! But you should mark the qrcode as the check-in qrcode of the operator, making it different from another similar image. For example it could have a red border and a specific text...

this is a great suggestion, thanks! Could you please open another task for this? Thank you!

**Comment by pleinad** [comment]
Hi Celestino. In the qrcode of the operator you store the password in clear, also now. My idea is to store the qrcode image in the application database with two-way encryption using a secret known only by the system about the user. I image that, even if the code is public, the hash of the password is stored securely in the database of specific instance of the system. I don't know if it is possible but, maybe, you can use this secret as base to build the key of the two-way ecnryption alghoritm.

For the red bordered qrcode, are you asking me to open a specific issue here in github?

Thank you, 
Daniel.

**Comment by cbellone** [comment]
> For the red bordered qrcode, are you asking me to open a specific issue here in github?

yes, please

thanks
Celestino
